### PR TITLE
DepClean Gradle Plugin implementation. (Not for merge)

### PR DESCRIPTION
**Pull request by ABHAY0O7**
Currently, I think I have implemented the core logic correctly, but still, there are some drawbacks that I am still trying to improve.
Also, the code that I have written until now is very unmanaged, there are lots of unnecessary commented out code, there are lots of method that I used for debugging and I provided very few java docs, so sorry for that. Please ignore these parts of code.
Currently to test the plugin,

1. First, Run `gradle clean build` without changing anything,
2. Run `gradle publishToMavenLocal`
3. Now uncomment the buildscript and apply plugin part that I have commented in build.gradle and again run `gradle clean build`
4. Run `gradle copyDependenciesLocally` task.
5. Run final `gradle debloat` task and you can see the results.

Thank you

### CODE REVIEW

**Comment by ABHAY0O7** [comment]
Hello @tdurieux @bbaudry 
I was thinking that if we could have a meet this weekend or may after that so that I can discuss the current status of the plugin, and also some problems that I am facing right now. The problems are quite complicated which I think can only be illustrated better at a meet, so it will be better for me.

**Review by tdurieux — CHANGES_REQUESTED** [review]
Could you remove all the Gradle folders, we would like to avoid to have .ar files inside the project?

Could you remove all the .gradle folder from the test resources?

Could you rename the folder depclean-gradle-plugin/src/Test to test?

It is for the moment difficult to review the changes when so many files are included in the PR. After you remove the folders I will continue to review the PR

**Comment by ABHAY0O7** [comment]
> Could you remove all the .gradle folder from the test resources?

I already removed them but I forgot to remove their caches, but now I removed them too.

> Could you rename the folder depclean-gradle-plugin/src/Test to test?

Renamed 👍🏽 

> Could you remove all the Gradle folders, we would like to avoid to have .ar files inside the project?

Are you talking about gradle.wrapper folders? If no, then please can you be more specific but If yes then in my knowledge we should not involve these folders in .gitignore as they are important ones. But if you want these folders to be ignored for now then I can remove them too. So please can you be more specific about this question that what I have to do.

File: depclean-gradle-plugin/build.gradle
Comment (tdurieux) [inline depclean-gradle-plugin/build.gradle:1]: Can you remove the comments?

File: depclean-gradle-plugin/gradle/wrapper/gradle-wrapper.properties
Comment (tdurieux) [inline depclean-gradle-plugin/gradle/wrapper/gradle-wrapper.properties:1]: Do you have to keep the wrapper folder?

File: depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleTask.java
Comment (tdurieux) [inline depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleTask.java:14]: The name of the function is not related to its body

File: depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleTask.java
Comment (tdurieux) [inline depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleTask.java:20]: Can be removed or should be improved.

File: depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/GradleProjectDependencyAnalysis.java
Comment (tdurieux) [inline depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/GradleProjectDependencyAnalysis.java:27]: to remove

File: depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/GradleProjectDependencyAnalysis.java
Comment (tdurieux) [inline depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/GradleProjectDependencyAnalysis.java:57]: Will it be used?

File: depclean-gradle-plugin/src/test/resources/all_dependencies_used/gradle/wrapper/gradle-wrapper.properties
Comment (tdurieux) [inline depclean-gradle-plugin/src/test/resources/all_dependencies_used/gradle/wrapper/gradle-wrapper.properties:1]: remove wrapper folder.

File: depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleAction.java
Comment (tdurieux) [inline depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleAction.java:51]: This method is too big, could you refactor it and split it into smaller entities (to improve maintainability)?

File: depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/DefaultGradleProjectDependencyAnalyzer.java
Comment (tdurieux) [inline depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/DefaultGradleProjectDependencyAnalyzer.java:48]: Could you create the same parameters that we have in the maven plugins? (could be done in a separate PR)

**Review by tdurieux — CHANGES_REQUESTED** [review]
Good job, the PR seems good. There are few elements that need to improve but it should be relatively fast.

File: depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/GradleProjectDependencyAnalysis.java
Comment (ABHAY0O7) [reply depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/GradleProjectDependencyAnalysis.java:27]: Are you talking about that comment here?

File: depclean-gradle-plugin/src/main/java/se/kth/depclean/depcleanGradleAction.java
Comment (ABHAY0O7) [reply depclean-gradle-plugin/src/main/java/se/kth/depclean/depcleanGradleAction.java]: Ok.

File: depclean-gradle-plugin/build.gradle
Comment (ABHAY0O7) [reply depclean-gradle-plugin/build.gradle:1]: Ok.

File: depclean-gradle-plugin/gradle/wrapper/gradle-wrapper.properties
Comment (ABHAY0O7) [reply depclean-gradle-plugin/gradle/wrapper/gradle-wrapper.properties:1]: Yes, it will be better if we keep the wrapper folder as it is very useful. In case if someone doesn't have Gradle installed in their system or is unupdated with the latest version then this wrapper folder will help them to build the project without any failure. For more information visit [here](https://stackoverflow.com/a/20351077/15557529)

File: depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleTask.java
Comment (ABHAY0O7) [reply depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleTask.java:14]: Can it be analyzeProject() or analyzeGradleProject()?

File: depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleTask.java
Comment (ABHAY0O7) [reply depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleTask.java:20]: Better remove.

File: depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/GradleProjectDependencyAnalysis.java
Comment (ABHAY0O7) [reply depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/GradleProjectDependencyAnalysis.java:57]: No, it will be of no use as I ignore scopes using the coordinates of the artifacts and not with any Artifact object itself.

Comment (ABHAY0O7) [reply depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/GradleProjectDependencyAnalysis.java:57]: So, I will remove it.

File: depclean-gradle-plugin/src/test/resources/all_dependencies_used/gradle/wrapper/gradle-wrapper.properties
Comment (ABHAY0O7) [reply depclean-gradle-plugin/src/test/resources/all_dependencies_used/gradle/wrapper/gradle-wrapper.properties:1]: Ya, I think we can remove these resources wrapper folders.

File: depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleAction.java
Comment (ABHAY0O7) [reply depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleAction.java:51]: Yes, I will try to split it as more as possible.

File: depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/DefaultGradleProjectDependencyAnalyzer.java
Comment (ABHAY0O7) [reply depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/DefaultGradleProjectDependencyAnalyzer.java:48]: I added all the parameters that we have in the maven plugin except for 3 parameters about which I talked about [here](https://github.com/castor-software/depclean/issues/92#issuecomment-873566633)
Please take a look there and then reply accordingly.

**Comment by ABHAY0O7** [comment]
Hello @tdurieux 
I have implemented all the requested changes except for those 2 which are still unresolved.
To resolve them, you have to be more clear that what I have to do.

Thank you.

File: depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/DefaultGradleProjectDependencyAnalyzer.java
Comment (tdurieux) [reply depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/DefaultGradleProjectDependencyAnalyzer.java:49]: I would still implement those parameters, we can have set the default to false for those additional parameters but those are useful.

File: depclean-gradle-plugin/build.gradle
Comment (tdurieux) [reply depclean-gradle-plugin/build.gradle]: you uncommented, you should remove those lines. if the plugin is not present on the system, it is not possible to compile it.

Comment (ABHAY0O7) [reply depclean-gradle-plugin/build.gradle]: Sorry that was my mistake.
I will correct it in my next push.

File: depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleAction.java
Comment (tdurieux) [inline depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleAction.java:102]: You could use Paths.get: https://docs.oracle.com/javase/7/docs/api/java/nio/file/Paths.html#get(java.lang.String,%20java.lang.String...)

File: depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/DefaultGradleProjectDependencyAnalyzer.java
Comment (tdurieux) [inline depclean-gradle-plugin/src/main/java/se/kth/depclean/analysis/DefaultGradleProjectDependencyAnalyzer.java:193]: You could use Paths.get: https://docs.oracle.com/javase/7/docs/api/java/nio/file/Paths.html#get(java.lang.String,%20java.lang.String...)

**Comment by tdurieux** [comment]
Looks good. `depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleAction.java` is really big maybe we should work on a way to make it smaller and easier to maintain.

File: depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleAction.java
Comment (ABHAY0O7) [reply depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleAction.java:102]: Thanks. that's a nice refactoring. I will implement it (wherever possible) in my next push.

**Comment by ABHAY0O7** [comment]
> Looks good. `depclean-gradle-plugin/src/main/java/se/kth/depclean/DepCleanGradleAction.java` is really big maybe we should work on a way to make it smaller and easier to maintain.

Hii @tdurieux,
As you suggested I reduced the size of the `DepCleanGradleAction`, to do this I introduce two new utility classes.
I also used the `Path` object instead of `String` path which saves a little bit extra lines of code.

**Comment by tdurieux** [comment]
I just realized that Github Action does not execute the test from the gradle plugin. Could you update https://github.com/castor-software/depclean/blob/master/.github/workflows/build.yml to include the gradle plugin tests. 
I think it is the final task before I can merge the PR.

It is a good job, continue like this!

**Comment by ABHAY0O7** [comment]
Hii @tdurieux 
I want you to please don't merge this PR for now.
This is PR is very messy and there are some useless commits also, so to make the history very clean and easily understandable, give me some time, I will update you soon.

Thank you

**Comment by ABHAY0O7** [comment]
Hii @tdurieux 
I have started to make the git history clean and understandable.
I have opened some PR's with their order of merge in their titles, now the commits are more clear and understandable with some description. 
I want you to review and merge the PRs as soon as possible, so that when I push new commits to my branches then they should be on new PR. :))

Thank you
Abhay

**Comment by ABHAY0O7** [comment]
Hii @tdurieux 
Nice to see you :))
I will update you with remaining PRs up to 7 PM GMT.

Thank you

**Comment by ABHAY0O7** [comment]
The work of this PR is divided and implemented int these PRs :

- #94 
- #95 
- #96 
- #97 
- #98 
- #99

**Pull request closed without merging**
